sycl/test/basic_tests/fp-accuracy.cpp

```diff
@@ -1,18 +1,56 @@
-// RUN: %clangxx -%fsycl-host-only -c -ffp-accuracy=high -faltmathlib=SVMLAltMathLibrary -fno-math-errno %s
+// Checks that the test is passing.
+// RUN: %clangxx -%fsycl-host-only -c -ffp-accuracy=high \
+// RUN: -faltmathlib=SVMLAltMathLibrary -fno-math-errno %s
+//
+// Checks that the attribute 'builtin-max-error' is generated.
+// RUN: %clangxx -c -fsycl -ffp-accuracy=high -fno-math-errno \
+// RUN: -S -emit-llvm -o - %s | FileCheck %s
 
 #include <sycl/sycl.hpp>
 using namespace sycl;
+constexpr access::mode sycl_write = access::mode::write;
 
 int main() {
   queue deviceQueue;
   double Value = 5.;
 
+  float input;
+
+  range<1> Length{1};
+
+  buffer<float, 1> out(1);
+
   deviceQueue.submit([&](handler &cgh) {
     cgh.single_task<class Kernel0>([=]() { double res = std::sin(Value); });
   });
 
   deviceQueue.submit([&](handler &cgh) {
     cgh.single_task<class Kernel1>([=]() { double res = sycl::sin(Value); });
   });
+
+  deviceQueue.submit([&](handler &cgh) {
+    sycl::accessor output{out, cgh, sycl::write_only};
+
+    cgh.single_task<class Kernel2>([=]() {
+      for (int i = 0; i < 1; i++)
+        output[i] = sycl::sin(input);
+    });
+  });
+  deviceQueue.submit([&](handler &cgh) {
+    auto output = out.template get_access<sycl_write>(cgh);
+
+    cgh.single_task<class Kernel3>([=]() {
+      for (int i = 0; i < 1; i++)
+        output[i] = sycl::sqrt(input);
+    });
+  });
+
+  // CHECK-LABEL: define {{.*}}spir_kernel void {{.*}}Kernel2
+  // CHECK: tail call {{.*}} float @llvm.fpbuiltin.sin.f32(float {{.*}}) #[[ATTR_HIGH:[0-9]+]]
+
+  // CHECK-LABEL: define {{.*}}spir_kernel void {{.*}}Kernel3
+  // CHECK: tail call noundef float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_HIGH]]
+
+  // CHECK: attributes #[[ATTR_HIGH]] = {{.*}}"fpbuiltin-max-error"="1.0"
```

Comment:
Does this patch affect device compilation?

Reply:
I'm not sure I understand your question. Different calls will be generated in the IR, so depending on the required accuracy it should result in a different implementation being selected. By the time we get to device compilation, these intrinsics will have been translated back to the __ocl_spv_* functions, but they should have an attribute attached with the accuracy requirement.

Reply:
I don't see any CHECKs for generated device code in this PR. We also have a FAST_MATH-based "optimization" implemented purely in our SYCL Device headers. I'd like to see a unit test that shows how this PR and the pre-existing optimization interact. Maybe we can/should drop the latter here as well.

Reply:
The CHECK lines in fp-accuracy.cpp represent IR generated from the device compilation (first compilation).  That's what you want right?
Can you point me to a unit test that uses the fast-math based optimizations?

Reply:
@aelovikov-intel Updated the run line as we discussed. Let me know if this works for you. Working on the -O2 version.

Reply:
> We also have a FAST_MATH-based "optimization" implemented purely in our SYCL Device headers.
Can you give me more information about this? I'm not sure what it means.

Reply:
It's done inside https://github.com/intel/llvm/blob/sycl/sycl/source/builtins_generator.py (although I'm in the process of major refactoring in this area). In a nutshell, we're doing this (that is what the script generates):

```
template <typename T0>
std::enable_if_t<(detail::is_vec_or_swizzle_v<T0> &&
                  detail::is_valid_elem_type_v<T0, float, double, half> &&
                  detail::is_valid_size_v<T0, 16, 1, 2, 3, 4, 8>),
                 detail::simplify_if_swizzle_t<T0>>
sin(T0 a0) __NOEXC {
  if constexpr (detail::use_fast_math_v<T0>) {
    return __sycl_std::__invoke_native_sin<detail::simplify_if_swizzle_t<T0>>(
        detail::simplify_if_swizzle_t<T0>(a0));
  }
  return __sycl_std::__invoke_sin<detail::simplify_if_swizzle_t<T0>>(
      detail::simplify_if_swizzle_t<T0>(a0));
}
```

where `use_fast_math_v` is defined like this https://github.com/intel/llvm/blob/fab8e8ce7acc86aba4cc26bfc1775ade2c6ab1d8/sycl/include/sycl/builtins_utils_scalar.hpp#L35-L42

Reply:
Second example: https://godbolt.org/z/hdcqT7K5G

That one shows the symbol not being defined, even though most fast-math flags will be set. The reason this is important is that we're changing the behavior of "-fp-model fast" to honor NaNs and infinites, so __FAST_MATH__ won't be defined.

Reply:
Compiling with -ffast-math actually doesn't create any fpbuiltin intrinsics. The reason being that the IR generated with 
-ffast-math for math functions is of the form __spirv_ocl_native_* and will therefore not go through this implementation. So this patch has no effect on the optimizations.

Reply:
> Compiling with -ffast-math actually doesn't create any fpbuiltin intrinsics. The reason being that the IR generated with
> -ffast-math for math functions is of the form __spirv_ocl_native_* and will therefore not go through this implementation. So this patch has no effect on the optimizations (at least for now, if we decide to also add these here).

I think that's a problem. That's probably happening because of the header file handling that Andrei mentioned, but I would assume that if the user specifies both -ffast-math and -ffp-accuracy=high, they want the high accuracy implementations. If they use -ffast-math and -ffp-accuracy=low, they probably want the native implementations where they exist, but we should be able to make that happen during the function selection phase. That's not happening now, but it should when that's faster than a library call and meets the accuracy requirements.

For CPU I came across this with sqrt in my testing. The __svml_sqrt1_ep function isn't as accurate as it claims to be, but the Numerics team confirmed that we should be using the native instruction in that case anyway because it's faster.

Reply:
> though in the near term that won't happen based on the -ffast-math option alone. It will take some time to get optimization working with the fpbuiltin intrinsics.

In this case, I think we should keep our header-based optimization in place to avoid performance regressions. I do think that we need some better testing here, but I wouldn't object on not handling that in this PR.